adr/0003-PlacementNew.md

```diff
@@ -0,0 +1,110 @@
+# C++ Data Lifetime Management via Placement New
+
+## Status
+PROPOSED
+
+Proposed by: Carlos Reyes (February 18, 2022)
+
+Discussed with:
+
+* Benson Ma (early ideas)
+* Eduardo Gonzalez (overall design goals)
+
+## Context
+Our architecture has a low-level C/C++ layer and a high level Java/Scala layer. We use JavaCpp to communicate between
+the layers. We would like to only use C++ code in the low level layer.
+
+Right now we allocate return values using C++ and deallocate them from Scala.
+This requires us to use C-style calloc/free system calls. We would like to migrate away from this design.
+
+## Proposal
+Eventually we want to move the return value deallocation to the same layer where the data is allocated,
+which is the low level layer. The safest way to do this is to rely on C++ destructors to clean up the heap-allocated
+objects. I view this proposal as an interim solution along the way to that goal.
+
+The idea is to maintain the C malloc/free calls while layering on top the C++ new/delete semantics.
+That is what the placement new syntax gives you. Please see the attached sample code.
```

Comment:
Like my previous comment, I think the proposal should state that using this syntax enables us to be able to use constructors/destructors in our C++ code which would unify things between Scala and C++ unit tests.